pkg/roles/repository.go

```diff
@@ -0,0 +1,273 @@
+// Copyright 2025 Canonical Ltd.
+// SPDX-License-Identifier: AGPL-3.0
+
+package roles
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+
+	sq "github.com/Masterminds/squirrel"
+
+	"github.com/canonical/identity-platform-admin-ui/internal/logging"
+	"github.com/canonical/identity-platform-admin-ui/internal/monitoring"
+	"github.com/canonical/identity-platform-admin-ui/internal/tracing"
+	"github.com/canonical/identity-platform-admin-ui/pkg/storage"
+)
+
+type RoleRepository struct {
+	db storage.DBClientInterface
+
+	logger  logging.LoggerInterface
+	tracer  tracing.TracingInterface
+	monitor monitoring.MonitorInterface
+}
+
+func (r *RoleRepository) FindRoleById(ctx context.Context, ID string) (*Role, error) {
+	ctx, span := r.tracer.Start(ctx, "roles.RoleRepository.FindRoleById")
+	defer span.End()
+
+	row := r.db.Statement().Select("id", "name", "owner").
+		From("role").
+		Where(sq.Eq{"id": ID}).
+		QueryRowContext(ctx)
+
+	var id, name, owner string
+
+	err := row.Scan(&id, &name, &owner)
+	if errors.Is(err, sql.ErrNoRows) {
+		return nil, storage.ErrNotFound
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("unable to scan FindRoleById result, %w", err)
+	}
+
+	return &Role{
+		ID:    id,
+		Name:  name,
+		Owner: owner,
+	}, nil
+}
+
+func (r *RoleRepository) FindRoleByIdAndOwner(ctx context.Context, ID string, userID string) (*Role, error) {
+	ctx, span := r.tracer.Start(ctx, "roles.RoleRepository.FindRoleByIdAndOwner")
+	defer span.End()
+
+	row := r.db.Statement().Select("id", "name", "owner").
+		From("role").
+		Where(sq.Eq{"id": ID}).
+		Where(sq.Eq{"owner": userID}).
+		QueryRowContext(ctx)
+
+	var id, name, owner string
+
+	err := row.Scan(&id, &name, &owner)
+	if errors.Is(err, sql.ErrNoRows) {
+		return nil, storage.ErrNotFound
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("unable to scan FindRoleByIdAndOwner result, %w", err)
+	}
+
+	return &Role{
+		ID:    id,
+		Name:  name,
+		Owner: owner,
+	}, nil
+}
+
+func (r *RoleRepository) FindRoleByNameAndOwner(ctx context.Context, name string, userID string) (*Role, error) {
```

Comment:
any reason to not use a variadic arg with the option pattern, something like withName or withOwner?

nice to have potentially, not a deal breaker

Reply:
it would be nice but since either 1 or 2 methods of these will be most likely removed after it wouldn't be worth it